lib/rouge/lexers/ecl.rb

```diff
@@ -0,0 +1,136 @@
+# -*- codding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class ECL < RegexLexer
+      tag 'ecl'
+      filenames '*.ecl'
+      mimetypes 'application/x-ecl'
+
+      title "ECL"
+      desc "Enterprise Control Language (hpccsystems.com)"
+      
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+
+      template = %w(
```

Comment:
We generally encourage the memoisation of these attributes. You can see an example of doing that in the JavaScript lexer:

https://github.com/rouge-ruby/rouge/blob/39b0ed7117a421fea1cd1403e8a4e17844653bc7/lib/rouge/lexers/javascript.rb#L99-L105

Reply:
I tried that in the very beginning but could not get it to work. I will try once again.

Reply:
The name of the method is written differently if it's within a block passed to `rule` (as was the case in the JavaScript example) than if it's within the rule itself. If it's in a block, you write it as `self.class.<method_name>` but if it's within the rule (as you currently use the local variables), you need to write it as `<LexerClass>.<method_name>`. If

Reply:
Appreciated!

Reply:
@pyrmont I'm having problem with case. From my various tests, it seems like the "include" function in memoisation's `"xxx.include? m[0]"` is  is case sensitive. ECL is not case sensitive. The reason I didn't use memoisation is that I don't want to put both cases into the list. Do you know of a way to make "include" in this case not case sensitive? Is there something on the `%w(` that can be applied? I'm not finding anything.

Reply:
In the block do this:

```ruby 
name = m[0].downcase
if self.class.keywords.include? name
  # ...
elsif self.class.others.include? name
  # ...
end
```

Reply:
Didn't find that anywhere. I will try that.

Reply:
It looks like you've memoised most of the variables except for `id`, `template` and `typed`. I understand why `id` would remain but is there any reason for `template` and `typed`?

Reply:
The templates start with "#".

Reply:
@pyrmont, are there still any pending changes necessary here?

Reply:
The `#` that's in regex pattern in the template rule is part of Ruby's interpolation syntax. If templates are supposed to start with a `#` then you need an additional `#` to start off the rule, like this:

```ruby
rule %r(#\b(#{template.join('|')})\b)i, Keyword::Type
```

Reply:
Good catch! Somehow I lost that change!

Reply:
@pyrmont fixed.